### ISSUE TITLE
LDAP SASL `EXTERNAL` broken in 4.2.0

### ISSUE THREAD

**Issue by thernstig** [ldapts/ldapts#133]
Version 4.1.1 works fine, but 4.2.0 breaks this:

```js
await client.bind('EXTERNAL');
```

It gives:

```sh
SaslBindInProgressError: SASL(0): successful result:  Code: 0xe
       at StatusCodeParser.parse (/home/user1/code/projectA/node_modules/ldapts/StatusCodeParser.js:28:24)
       at Client._sendBind (/home/user1/code/projectA/node_modules/ldapts/Client.js:472:55)
       at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
       at async Client.bindSASL (/home/user1/code/projectA/node_modules/ldapts/Client.js:168:9)
       at async Client.bind (/home/user1/code/projectA/node_modules/ldapts/Client.js:141:13)
       at async getUserDetails (file:///home/user1/code/projectA/framework/authentication/ldapAdp.js:210:7)
       at async Strategy._verify (file:///home/user1/code/projectA/framework/authentication/ldapAdp.js:100:13) {
     code: 14,
     response: BindResponse {
       version: 3,
       messageId: 2,
       controls: [],
       status: 14,
       matchedDN: '',
       errorMessage: 'SASL(0): successful result: ',
       data: [Array],
       protocolOperation: 97
     }
   }
 }
```

There is something in https://github.com/ldapts/ldapts/pull/120 that must handle this incorrect.

For info: Calling BIND twice does not help.

cc @jgeurts @TimoHocker

**Comment by TimoHocker** [comment]
it's weird that the message says successful while the status code is 14 (bind in progress).
It would be necessary to find out if there is some bug that causes it to give the wrong status code or if the server just responds weird, then it would make sense to just declare that specific response as successful.
I don't have a way to test the external bind though, I probably won't be able to help much.

**Comment by thernstig** [comment]
@TimoHocker since this is a regression, I'd then rather your commit was reverted since it broke existing functionality.

Maybe @wattry can look at the commit you made as well, since he helped @jgeurts to fix the initial SASL EXTERNAL support in https://github.com/ldapts/ldapts/issues/89.

**Comment by TimoHocker** [comment]
That would cause the same problem on the other side, it breaks the other SASL mechanisms for anyone using them. The best way to work on this issue would be to find out what is wrong and fix that with you @thernstig, @wattry and possibly others with access to a server that offers SASL EXTERNAL testing it. I don't think a regression should be fixed with a breaking change just because it's quick.

One solution would be to add a separate handler for EXTERNAL, that replicates the old function, maybe it's even enough to remove it from the list of SASL Mechanisms, I don't remember exactly how it was handled before. I'll look into it when I have time this week.

**Comment by TimoHocker** [comment]
From what I can tell there is no difference in the handling of EXTERNAL between 4.1.1 and 4.2.0. There is one small difference in the request that leaves out the password field if it's not a string, because null or undefined were marked as malformed.

My only guess right now is that your server doesn't like the password field gone. Can you please test with the commit `41da44f` reverted or the repo checked out at `e74d898`? If it's possible and the traffic is not encrypted, then try to check with wireshark if there are any differences between the working and non working version.

**Comment by TimoHocker** [comment]
if the empty password really is required for external binds to function then it would make sense to revert the commit `41da44f`.
I made that change in order to fix a warning the server gave me, but that's only a cosmetic thing.

**Comment by thernstig** [comment]
@wattry @TimoHocker the fix works :)

**Comment by jgeurts** [comment]
Could you please try ldapts@5.0.0-beta1 when you have a chance?

**Comment by TimoHocker** [comment]
my tests are successful too

**Comment by thernstig** [comment]
Mine too

**Comment by jgeurts** [comment]
This should be fixed with v5.0.0
